src/middlewared/middlewared/gdb.py

```diff
@@ -0,0 +1,190 @@
+
+    def get_dataset(self, name=None):
+        zfs = libzfs.ZFS()
+        if name is None:
+            for pool in zfs.pools:
+                if pool.name == 'freenas-boot':
+                    continue
+                return pool.root_dataset
+            raise RuntimeError('no dataset found')
+        else:
+            return zfs.get_dataset(name)
+
+    def _get_version(self):
+        with open('/etc/version', 'r') as f:
+            version = f.read().split()[0]
+
+        product, version = version.split('-', 1)
+        major = version.split('-')[0].split('.')[0]
+
+        return {
+            'major': major,
+            'product': product,
+            'version': version,
+        }
+
+    def debug_symbols_url(self):
+        version = self._get_version()
+        if version['product'] == 'FreeNAS':
+            base_url = f'https://download.freenas.org/{version["major"]}'
+            if 'MASTER' in version['version']:
+                verdir = f'MASTER/{version["version"].split("-")[-1]}'
+            else:
+                verdir = version['version']
+            dirurl = f'{base_url}/{verdir}/x64/'
+            req = requests.get(dirurl)
+            if req.status_code != 200:
+                raise RuntimeError('debug symbols url could not be found, provide the url')
+            reg = re.search(r'href="(?P<debug>.+?\.debug\.txz)"', req.text, re.M)
+            if reg is None:
+                raise RuntimeError(f'debug symbols file not found in {dirurl}')
+            return f'{dirurl}{reg.group("debug")}'
+        else:
+            raise RuntimeError(f'product {version["product"]} not supported')
+
+    def _download_debug(self, dataset, url):
+        filename = url.rsplit('/', 1)[-1]
+        path = f'{dataset.mountpoint}/{filename}'
+
+        if os.path.exists(path):
+            first_byte = os.path.getsize(path)
+        else:
+            first_byte = 0
+
+        headers = {}
+        r = requests.head(url)
+        total_size = int(r.headers.get('content-length', 0))
+        if total_size:
+            headers['Range'] = f'bytes={first_byte}-{total_size}'
+
+        if first_byte and first_byte == total_size:
+            return path
+
+        if first_byte > 0 and total_size == first_byte:
+            return path
+
+        r = requests.get(url, headers=headers, stream=True)
+        content_range = r.headers.get('content-range')
+        if content_range:
+            initial = int(content_range.split()[1].split('-')[0])
+        else:
+            initial = 0
+
+        chunk_size = 64 * 1024
+        with open(path, 'ab') as f:
+            if not initial:
+                f.seek(0)
+            pbar = tqdm.tqdm(total=total_size, initial=initial, unit='B', unit_scale=True)
+            for chunk in r.iter_content(chunk_size):
+                f.write(chunk)
+                pbar.update(chunk_size)
+        return path
+
+    def extract(self, dataset, path):
+        proc = subprocess.Popen([
```

Comment:
I think `subprocess.check_call` idiom is more convenient here: user can see what command failed, what was the output and exit code

Reply:
My only concern is this: ```Note Do not use stdout=PIPE or stderr=PIPE with this function. The child process will block if it generates enough output to a pipe to fill up the OS pipe buffer as the pipes are not being read from.```

What if the output happens to be big enough? (e.g. a lot of errors on stderr about permissions or somesuch)

Reply:
I suggest to use `None` instead of `PIPE`, in my opinion redirection is exactly what is needed here.